pkg/app/pipedv1/plugin/kubernetes/livestate/plugin_test.go

```diff
@@ -0,0 +1,378 @@
+func makeTestDiffChange(t *testing.T, oldYAML, newYAML string) provider.DiffListChange {
+	t.Helper()
+	old := makeTestManifest(t, oldYAML)
+	new := makeTestManifest(t, newYAML)
+
+	oldData, err := old.MarshalJSON()
+	require.NoError(t, err)
+	oldUnstructured := unstructured.Unstructured{}
+	err = oldUnstructured.UnmarshalJSON(oldData)
+	require.NoError(t, err)
+
+	newData, err := new.MarshalJSON()
+	require.NoError(t, err)
+	newUnstructured := unstructured.Unstructured{}
+	err = newUnstructured.UnmarshalJSON(newData)
+	require.NoError(t, err)
+
+	result, err := diff.DiffUnstructureds(oldUnstructured, newUnstructured, old.Key().String(),
+		diff.WithEquateEmpty(),
+		diff.WithIgnoreAddingMapKeys(),
+		diff.WithCompareNumberAndNumericString(),
+	)
+	require.NoError(t, err)
+
+	return provider.DiffListChange{
+		Old:  old,
+		New:  new,
+		Diff: result,
+	}
+}
+
+func TestCalculateSyncState(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name       string
+		diffResult *provider.DiffListResult
+		commitHash string
+		want       sdk.ApplicationSyncState
+	}{
+		{
+			name: "all resources are in sync",
+			diffResult: &provider.DiffListResult{
+				Adds:    []provider.Manifest{},
+				Deletes: []provider.Manifest{},
+				Changes: []provider.DiffListChange{},
+			},
+			commitHash: "1234567",
+			want: sdk.ApplicationSyncState{
+				Status:      sdk.ApplicationSyncStateSynced,
+				ShortReason: "",
+				Reason:      "",
+			},
+		},
+		{
+			name: "deployment spec changes",
+			diffResult: &provider.DiffListResult{
+				Changes: []provider.DiffListChange{
+					makeTestDiffChange(t, `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: test-deployment
+  namespace: default
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: test
+  template:
+    metadata:
+      labels:
+        app: test
+    spec:
+      containers:
+      - name: app
+        image: nginx:1.19
+`, `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: test-deployment
+  namespace: default
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app: test
+  template:
+    metadata:
+      labels:
+        app: test
+    spec:
+      containers:
+      - name: app
+        image: nginx:1.20
+`),
+				},
+			},
+			commitHash: "1234567",
+			want: sdk.ApplicationSyncState{
+				Status:      sdk.ApplicationSyncStateOutOfSync,
+				ShortReason: "There are 1 manifests not synced (0 adds, 0 deletes, 1 changes)",
+				Reason: `Diff between the defined state in Git at commit 1234567 and actual state in cluster:
+
+--- Actual   (LiveState)
++++ Expected (Git)
+
+# 1. name="test-deployment", kind="Deployment", namespace="default", apiGroup="apps"
+
+  spec:
+    #spec.replicas
+-   replicas: 1
++   replicas: 3
+
+    template:
+      spec:
+        containers:
+          -
+            #spec.template.spec.containers.0.image
+-           image: nginx:1.19
++           image: nginx:1.20
+
+
+`,
+			},
+		},
+		{
+			name: "service and ingress changes",
```

Comment:
[ask] Does the logic for `calculateSyncState` depend on each resource kind?

Reply:
The ConfigMap and Secrets are masked. The other kinds are treated as the same.

Reply:
I see. Then, we can fix the test case name to clarify the spec for the function.
For example "changed two resources". WDYT?

Reply:
OK, it's a good name to show the spec.
Also, I'll change the second test case, `deployment spec changes`, to `changed one resource`.

Reply:
Fixed on this commit
https://github.com/pipe-cd/pipecd/pull/5676/commits/9a4b84a645765f49c7573432d4830d038dd690f4